File: src/lib/routing.ts
import type { Coordinates, QuestStop, RouteLeg } from "./types";

const WALK_METERS_PER_MINUTE = 78;

export function getWalkingMinutes(from: QuestStop, to: QuestStop) {
  if (from.walkMinutesToNext) return from.walkMinutesToNext;
  return Math.max(4, Math.round(distanceMeters(from.coordinates, to.coordinates) / WALK_METERS_PER_MINUTE));
}

export function buildRouteLegs(stops: QuestStop[]): RouteLeg[] {
  return stops.slice(0, -1).map((from, index) => {
    const to = stops[index + 1];
    const minutes = getWalkingMinutes(from, to);
    return {
      fromStopId: from.id,
      toStopId: to.id,
      minutes,
      distanceMeters: Math.round(minutes * WALK_METERS_PER_MINUTE),
      polyline: interpolatePolyline(from.coordinates, to.coordinates)
    };
  });
}

export function routeTotals(stops: QuestStop[]) {
  const stopMinutes = stops.reduce((total, stop) => total + stop.visitMinutes, 0);
  const walkMinutes = buildRouteLegs(stops).reduce((total, leg) => total + leg.minutes, 0);
  return {
    stopMinutes,
    walkMinutes,
    totalMinutes: stopMinutes + walkMinutes
  };
}

export function distanceMeters(from: Coordinates, to: Coordinates) {
  const radius = 6371000;
  const lat1 = toRadians(from.lat);
  const lat2 = toRadians(to.lat);
  const deltaLat = toRadians(to.lat - from.lat);
  const deltaLng = toRadians(to.lng - from.lng);
  const a =
    Math.sin(deltaLat / 2) * Math.sin(deltaLat / 2) +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(deltaLng / 2) * Math.sin(deltaLng / 2);
  return 2 * radius * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function interpolatePolyline(from: Coordinates, to: Coordinates): Coordinates[] {
  const midpoint = {
    lat: (from.lat + to.lat) / 2 + 0.0005,
    lng: (from.lng + to.lng) / 2 - 0.0004
  };
  return [from, midpoint, to];
}

function toRadians(value: number) {
  return (value * Math.PI) / 180;
}

File: backend/app/routing.py
from __future__ import annotations

import math

from app.models import Coordinates, QuestStop, RouteLeg

WALK_METERS_PER_MINUTE = 78


def get_walking_minutes(source: QuestStop, target: QuestStop) -> int:
    if source.walk_minutes_to_next:
        return source.walk_minutes_to_next
    return max(4, round(distance_meters(source.coordinates, target.coordinates) / WALK_METERS_PER_MINUTE))


def build_route_legs(stops: list[QuestStop]) -> list[RouteLeg]:
    legs: list[RouteLeg] = []
    for index, source in enumerate(stops[:-1]):
        target = stops[index + 1]
        minutes = get_walking_minutes(source, target)
        legs.append(
            RouteLeg(
                fromStopId=source.id,
                toStopId=target.id,
                minutes=minutes,
                distanceMeters=round(minutes * WALK_METERS_PER_MINUTE),
                polyline=interpolate_polyline(source.coordinates, target.coordinates),
            )
        )
    return legs


def route_totals(stops: list[QuestStop]) -> tuple[int, int, int]:
    stop_minutes = sum(stop.visit_minutes for stop in stops)
    walk_minutes = sum(leg.minutes for leg in build_route_legs(stops))
    return stop_minutes, walk_minutes, stop_minutes + walk_minutes


def distance_meters(source: Coordinates, target: Coordinates) -> float:
    radius = 6_371_000
    lat1 = math.radians(source.lat)
    lat2 = math.radians(target.lat)
    delta_lat = math.radians(target.lat - source.lat)
    delta_lng = math.radians(target.lng - source.lng)
    a = (
        math.sin(delta_lat / 2) * math.sin(delta_lat / 2)
        + math.cos(lat1) * math.cos(lat2) * math.sin(delta_lng / 2) * math.sin(delta_lng / 2)
    )
    return 2 * radius * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def interpolate_polyline(source: Coordinates, target: Coordinates) -> list[Coordinates]:
    midpoint = Coordinates(
        lat=(source.lat + target.lat) / 2 + 0.0005,
        lng=(source.lng + target.lng) / 2 - 0.0004,
    )
    return [source, midpoint, target]

